cranelift/codegen/src/ir/instructions.rs

```diff
@@ -31,6 +31,83 @@ pub type ValueList = entity::EntityList<Value>;
 /// Memory pool for holding value lists. See `ValueList`.
 pub type ValueListPool = entity::ListPool<Value>;
 
+/// A pair of a Block and its arguments, stored in a single EntityList internally.
+///
+/// NOTE: We don't expose either value_to_block or block_to_value outside of this module because
+/// this operation is not generally safe. However, as the two share the same underlying layout,
+/// they can be stored in the same value pool.
+///
+/// BlockWithArgs makes use of this shared layout by storing all of its contents (a block and its
+/// argument) in a single EntityList. This is a bit better than introducing a new entity type for
+/// the pair of a block name and the arguments entity list, as we don't pay any indirection penalty
+/// to get to the argument values -- they're stored in-line with the block in the same list.
+///
+/// The BlockWithArgs::new function guarantees this layout by requiring a block argument that's
+/// written in as the first element of the EntityList. Any subsequent entries are always assumed to
+/// be real Values.
+#[derive(Debug, Clone, Copy, PartialEq, Hash)]
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+pub struct BlockWithArgs {
+    /// The underlying storage for the BlockWithArgs. The first element of the values EntityList is
+    /// guaranteed to always be a Block encoded as a Value via BlockWithArgs::block_to_value.
+    /// Consequently, the values entity list is never empty.
+    values: entity::EntityList<Value>,
+}
+
+impl BlockWithArgs {
+    // NOTE: the only uses of this function should be internal to BlockWithArgs. See the block
+    // comment on BlockWithArgs for more context.
+    fn value_to_block(val: Value) -> Block {
+        Block::from_u32(val.as_u32())
+    }
+
+    // NOTE: the only uses of this function should be internal to BlockWithArgs. See the block
+    // comment on BlockWithArgs for more context.
+    fn block_to_value(block: Block) -> Value {
+        Value::from_u32(block.as_u32())
+    }
+
+    /// Construct a BlockWithArgs with the given block and arguments.
+    pub fn new(block: Block, args: &[Value], pool: &mut ValueListPool) -> Self {
+        let mut values = ValueList::default();
+        values.push(Self::block_to_value(block), pool);
+        values.extend(args.iter().copied(), pool);
+        Self { values }
+    }
+
+    /// Return the block for this BlockWithArgs.
+    pub fn block(&self, pool: &ValueListPool) -> Block {
+        let val = self.values.first(pool).unwrap();
+        Self::value_to_block(val)
+    }
+
+    /// Replace the block for this BlockWithArgs.
+    pub fn set_block(&mut self, block: Block, pool: &mut ValueListPool) {
+        let cell = self.values.get_mut(0, pool).unwrap();
```

Comment:
Somewhat confusing to name this local `cell` if it's not actually a Cell or some other kind of interior mutability -- can we just call it `first_elem` or something?